File: go/bcenvelope/notation.go
package bcenvelope

import (
	"fmt"
	"math"
	"slices"
	"strings"

	dcbor "github.com/nickel-blockchaincommons/dcbor-go"
	knownvalues "github.com/nickel-blockchaincommons/knownvalues-go"
)

// envelopeFormatOpts controls envelope notation formatting.
type envelopeFormatOpts struct {
	flat    bool
	context FormatContextOpt
}

func defaultEnvelopeFormatOpts() envelopeFormatOpts {
	return envelopeFormatOpts{context: FormatOptGlobal()}
}

// envelopeFormatItem represents a formatted element in envelope notation.
// This is an internal type used by the notation formatting system.
type envelopeFormatItem struct {
	kind  formatItemKind
	str   string
	items []envelopeFormatItem
}

type formatItemKind int

const (
	formatItemBegin formatItemKind = iota
	formatItemEnd
	formatItemItem
	formatItemSeparator
	formatItemList
)

func formatBegin(s string) envelopeFormatItem {
	return envelopeFormatItem{kind: formatItemBegin, str: s}
}

func formatEnd(s string) envelopeFormatItem {
	return envelopeFormatItem{kind: formatItemEnd, str: s}
}

func formatItem(s string) envelopeFormatItem {
	return envelopeFormatItem{kind: formatItemItem, str: s}
}

func formatSeparator() envelopeFormatItem {
	return envelopeFormatItem{kind: formatItemSeparator}
}

func formatList(items []envelopeFormatItem) envelopeFormatItem {
	return envelopeFormatItem{kind: formatItemList, items: items}
}

func (item envelopeFormatItem) flatten() []envelopeFormatItem {
	if item.kind == formatItemList {
		var result []envelopeFormatItem
		for _, child := range item.items {
			result = append(result, child.flatten()...)
		}
		return result
	}
	return []envelopeFormatItem{item}
}

func nicenItems(items []envelopeFormatItem) []envelopeFormatItem {
	input := make([]envelopeFormatItem, len(items))
	copy(input, items)
	var result []envelopeFormatItem

	for len(input) > 0 {
		current := input[0]
		input = input[1:]
		if len(input) == 0 {
			result = append(result, current)
			break
		}
		if current.kind == formatItemEnd {
			if input[0].kind == formatItemBegin {
				result = append(result, formatEnd(current.str+" "+input[0].str))
				result = append(result, formatBegin(""))
				input = input[1:]
			} else {
				result = append(result, current)
			}
		} else {
			result = append(result, current)
		}
	}

	return result
}

func formatIndent(level int) string {
	return strings.Repeat(" ", level*4)
}

func addSpaceAtEndIfNeeded(s string) string {
	if s == "" {
		return " "
	}
	if strings.HasSuffix(s, " ") {
		return s
	}
	return s + " "
}

func (item envelopeFormatItem) format(opts envelopeFormatOpts) string {
	if opts.flat {
		return item.formatFlat()
	}
	return item.formatHierarchical()
}

func (item envelopeFormatItem) formatFlat() string {
	var line string
	items := item.flatten()
	for _, it := range items {
		switch it.kind {
		case formatItemBegin:
			if !strings.HasSuffix(line, " ") {
				line += " "
			}
			line += it.str
			line += " "
		case formatItemEnd:
			if !strings.HasSuffix(line, " ") {
				line += " "
			}
			line += it.str
			line += " "
		case formatItemItem:
			line += it.str
		case formatItemSeparator:
			line = strings.TrimRight(line, " ") + ", "
		case formatItemList:
			for _, child := range it.items {
				line += child.formatFlat()
			}
		}
	}
	return line
}

func (item envelopeFormatItem) formatHierarchical() string {
	var lines []string
	level := 0
	currentLine := ""
	items := nicenItems(item.flatten())
	for _, it := range items {
		switch it.kind {
		case formatItemBegin:
			if it.str != "" {
				c := it.str
				if currentLine != "" {
					c = addSpaceAtEndIfNeeded(currentLine) + it.str
				}
				lines = append(lines, formatIndent(level)+c+"\n")
			}
			level++
			currentLine = ""
		case formatItemEnd:
			if currentLine != "" {
				lines = append(lines, formatIndent(level)+currentLine+"\n")
				currentLine = ""
			}
			level--
			lines = append(lines, formatIndent(level)+it.str+"\n")
		case formatItemItem:
			currentLine += it.str
		case formatItemSeparator:
			if currentLine != "" {
				lines = append(lines, formatIndent(level)+currentLine+"\n")
				currentLine = ""
			}
		case formatItemList:
			lines = append(lines, "<list>")
		}
	}
	if currentLine != "" {
		lines = append(lines, currentLine)
	}
	return strings.Join(lines, "")
}

// formatItemIndex returns a sort index for the item kind.
func (item envelopeFormatItem) formatItemIndex() int {
	switch item.kind {
	case formatItemBegin:
		return 1
	case formatItemEnd:
		return 2
	case formatItemItem:
		return 3
	case formatItemSeparator:
		return 4
	case formatItemList:
		return 5
	default:
		return 0
	}
}

// compareFormatItems compares two format items for sorting.
func compareFormatItems(a, b envelopeFormatItem) int {
	ai := a.formatItemIndex()
	bi := b.formatItemIndex()
	if ai < bi {
		return -1
	}
	if ai > bi {
		return 1
	}
	if a.str < b.str {
		return -1
	}
	if a.str > b.str {
		return 1
	}
	// For list items, compare child items recursively
	if a.kind == formatItemList && b.kind == formatItemList {
		minLen := len(a.items)
		if len(b.items) < minLen {
			minLen = len(b.items)
		}
		for i := 0; i < minLen; i++ {
			cmp := compareFormatItems(a.items[i], b.items[i])
			if cmp != 0 {
				return cmp
			}
		}
		if len(a.items) < len(b.items) {
			return -1
		}
		if len(a.items) > len(b.items) {
			return 1
		}
	}
	return 0
}

// Envelope formatting methods.

// Format returns the envelope notation for this envelope using the global format context.
func (e *Envelope) Format() string {
	return e.FormatOpt(defaultEnvelopeFormatOpts())
}

// FormatFlat returns the envelope notation in flat (single-line) format.
func (e *Envelope) FormatFlat() string {
	opts := defaultEnvelopeFormatOpts()
	opts.flat = true
	return e.FormatOpt(opts)
}

// FormatOpt returns the envelope notation with the given options.
func (e *Envelope) FormatOpt(opts envelopeFormatOpts) string {
	item := e.formatItem(opts)
	return strings.TrimSpace(item.format(opts))
}

// formatItem returns an envelopeFormatItem for this envelope's notation.
func (e *Envelope) formatItem(opts envelopeFormatOpts) envelopeFormatItem {
	switch c := e.Case().(type) {
	case *LeafCase:
		return formatCBOR(*c.CBOR, opts)
	case *WrappedCase:
		return formatList([]envelopeFormatItem{
			formatBegin("{"),
			c.Envelope.formatItem(opts),
			formatEnd("}"),
		})
	case *AssertionCase:
		return c.Assertion.formatItem(opts)
	case *KnownValueCase:
		return formatKnownValue(c.Value, opts)
	case *EncryptedCase:
		return formatItem("ENCRYPTED")
	case *CompressedCase:
		return formatItem("COMPRESSED")
	case *NodeCase:
		return formatNode(c, opts)
	case *ElidedCase:
		return formatItem("ELIDED")
	default:
		return formatItem("<error>")
	}
}

func formatNode(nc *NodeCase, opts envelopeFormatOpts) envelopeFormatItem {
	var items []envelopeFormatItem

	subjectItem := nc.Subject.formatItem(opts)
	var elidedCount int
	var encryptedCount int
	var compressedCount int
	var typeAssertionItems [][]envelopeFormatItem
	var assertionItems [][]envelopeFormatItem

	for _, assertion := range nc.Assertions {
		switch assertion.Case().(type) {
		case *ElidedCase:
			elidedCount++
		case *EncryptedCase:
			encryptedCount++
		case *CompressedCase:
			compressedCount++
		default:
			item := []envelopeFormatItem{assertion.formatItem(opts)}
			// Check if this is a type assertion (isA predicate)
			isTypeAssertion := false
			if pred := assertion.AsPredicate(); pred != nil {
				if kv := pred.Subject().AsKnownValue(); kv != nil {
					if kv.Equal(knownvalues.IsA) {
						isTypeAssertion = true
					}
				}
			}
			if isTypeAssertion {
				typeAssertionItems = append(typeAssertionItems, item)
			} else {
				assertionItems = append(assertionItems, item)
			}
		}
	}

	// Sort type assertion items
	sortFormatItemGroups(typeAssertionItems)
	// Sort assertion items
	sortFormatItemGroups(assertionItems)
	// Prepend type assertions
	assertionItems = append(typeAssertionItems, assertionItems...)

	// Add compressed markers
	if compressedCount > 1 {
		assertionItems = append(assertionItems, []envelopeFormatItem{
			formatItem(fmt.Sprintf("COMPRESSED (%d)", compressedCount)),
		})
	} else if compressedCount > 0 {
		assertionItems = append(assertionItems, []envelopeFormatItem{
			formatItem("COMPRESSED"),
		})
	}

	// Add elided markers
	if elidedCount > 1 {
		assertionItems = append(assertionItems, []envelopeFormatItem{
			formatItem(fmt.Sprintf("ELIDED (%d)", elidedCount)),
		})
	} else if elidedCount > 0 {
		assertionItems = append(assertionItems, []envelopeFormatItem{
			formatItem("ELIDED"),
		})
	}

	// Add encrypted markers
	if encryptedCount > 1 {
		assertionItems = append(assertionItems, []envelopeFormatItem{
			formatItem(fmt.Sprintf("ENCRYPTED (%d)", encryptedCount)),
		})
	} else if encryptedCount > 0 {
		assertionItems = append(assertionItems, []envelopeFormatItem{
			formatItem("ENCRYPTED"),
		})
	}

	// Intersperse separators
	var joinedItems []envelopeFormatItem
	for i, group := range assertionItems {
		if i > 0 {
			joinedItems = append(joinedItems, formatSeparator())
		}
		joinedItems = append(joinedItems, group...)
	}

	needsBraces := nc.Subject.IsSubjectAssertion()

	if needsBraces {
		items = append(items, formatBegin("{"))
	}
	items = append(items, subjectItem)
	if needsBraces {
		items = append(items, formatEnd("}"))
	}
	items = append(items, formatBegin("["))
	items = append(items, joinedItems...)
	items = append(items, formatEnd("]"))
	return formatList(items)
}

func sortFormatItemGroups(groups [][]envelopeFormatItem) {
	slices.SortStableFunc(groups, func(a, b []envelopeFormatItem) int {
		minLen := len(a)
		if len(b) < minLen {
			minLen = len(b)
		}
		for k := 0; k < minLen; k++ {
			cmp := compareFormatItems(a[k], b[k])
			if cmp != 0 {
				return cmp
			}
		}
		if len(a) < len(b) {
			return -1
		}
		if len(a) > len(b) {
			return 1
		}
		return 0
	})
}

// formatCBOR formats a CBOR value for envelope notation.
func formatCBOR(cbor dcbor.CBOR, opts envelopeFormatOpts) envelopeFormatItem {
	// Check if this is a tagged envelope
	if tag, inner, ok := cbor.AsTaggedValue(); ok {
		envelopeTags := EnvelopeCBORTags()
		if len(envelopeTags) > 0 && tag.Value() == envelopeTags[0].Value() {
			env, err := EnvelopeFromUntaggedCBOR(inner)
			if err == nil {
				return env.formatItem(opts)
			}
			return formatItem("<error>")
		}
	}
	summary, err := envelopeSummaryCBOR(cbor, math.MaxInt, opts.context)
	if err != nil {
		return formatItem("<error>")
	}
	return formatItem(summary)
}

// formatKnownValue formats a KnownValue for envelope notation.
func formatKnownValue(kv knownvalues.KnownValue, opts envelopeFormatOpts) envelopeFormatItem {
	var name string
	switch opts.context.Mode {
	case FormatContextNone:
		name = kv.Name()
	case FormatContextGlobal:
		name = WithFormatContext(func(ctx *FormatContext) string {
			if assigned, ok := ctx.KnownValues().AssignedName(kv); ok {
				return assigned
			}
			return kv.Name()
		})
	case FormatContextCustom:
		if opts.context.Context != nil {
			if assigned, ok := opts.context.Context.KnownValues().AssignedName(kv); ok {
				name = assigned
			} else {
				name = kv.Name()
			}
		} else {
			name = kv.Name()
		}
	}
	return formatItem("'" + name + "'")
}

// Assertion formatting.
func (a *Assertion) formatItem(opts envelopeFormatOpts) envelopeFormatItem {
	return formatList([]envelopeFormatItem{
		a.Predicate().formatItem(opts),
		formatItem(": "),
		a.Object().formatItem(opts),
	})
}

// Envelope String returns a debug description.
func (e *Envelope) String() string {
	opts := defaultEnvelopeFormatOpts()
	switch c := e.Case().(type) {
	case *NodeCase:
		var assertionStrs []string
		for _, a := range c.Assertions {
			assertionStrs = append(assertionStrs, a.String())
		}
		assertions := "[" + strings.Join(assertionStrs, ", ") + "]"
		return fmt.Sprintf(".node(%s, %s)", c.Subject, assertions)
	case *LeafCase:
		return fmt.Sprintf(".cbor(%s)", formatCBOR(*c.CBOR, opts).str)
	case *WrappedCase:
		return fmt.Sprintf(".wrapped(%s)", c.Envelope)
	case *AssertionCase:
		return fmt.Sprintf(".assertion(%s, %s)", c.Assertion.Predicate(), c.Assertion.Object())
	case *ElidedCase:
		return ".elided"
	case *KnownValueCase:
		return fmt.Sprintf(".knownValue(%s)", c.Value)
	case *EncryptedCase:
		return ".encrypted"
	case *CompressedCase:
		return ".compressed"
	default:
		return "<unknown>"
	}
}

File: python/bc-envelope/src/bc_envelope/_notation.py
"""Internal notation rendering for envelope formatting.

Renders each envelope case as a formatted string with annotations.
Produces the human-readable "envelope notation" that shows the semantic
structure of envelopes.
"""

from __future__ import annotations

import enum
from typing import TYPE_CHECKING

from dcbor import CBORCase
from known_values import IS_A

from ._envelope_case import CaseType
from ._envelope_summary import cbor_envelope_summary
from ._format_context import (
    FormatContextOpt,
    flanked_by,
    with_format_context,
)

if TYPE_CHECKING:
    from dcbor import CBOR

    from ._assertion import Assertion
    from ._envelope import Envelope


# ---------------------------------------------------------------------------
# EnvelopeFormatItem -- represents one element of formatted output
# ---------------------------------------------------------------------------

class FormatItemKind(enum.IntEnum):
    BEGIN = 1
    END = 2
    ITEM = 3
    SEPARATOR = 4
    LIST = 5


class EnvelopeFormatItem:
    """Represents one element in the formatted envelope notation."""

    __slots__ = ("kind", "text", "items")

    def __init__(
        self,
        kind: FormatItemKind,
        text: str = "",
        items: list[EnvelopeFormatItem] | None = None,
    ) -> None:
        self.kind = kind
        self.text = text
        self.items: list[EnvelopeFormatItem] = items if items is not None else []

    @staticmethod
    def begin(delimiter: str) -> EnvelopeFormatItem:
        return EnvelopeFormatItem(FormatItemKind.BEGIN, delimiter)

    @staticmethod
    def end(delimiter: str) -> EnvelopeFormatItem:
        return EnvelopeFormatItem(FormatItemKind.END, delimiter)

    @staticmethod
    def item(text: str) -> EnvelopeFormatItem:
        return EnvelopeFormatItem(FormatItemKind.ITEM, text)

    @staticmethod
    def separator() -> EnvelopeFormatItem:
        return EnvelopeFormatItem(FormatItemKind.SEPARATOR)

    @staticmethod
    def list_(items: list[EnvelopeFormatItem]) -> EnvelopeFormatItem:
        return EnvelopeFormatItem(FormatItemKind.LIST, items=items)

    # -- Comparison for sorting assertions --

    def _sort_key(self) -> tuple:
        if self.kind == FormatItemKind.LIST:
            return (self.kind, "", tuple(i._sort_key() for i in self.items))
        return (self.kind, self.text, ())

    def __lt__(self, other: object) -> bool:
        if not isinstance(other, EnvelopeFormatItem):
            return NotImplemented
        return self._sort_key() < other._sort_key()

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, EnvelopeFormatItem):
            return NotImplemented
        return (
            self.kind == other.kind
            and self.text == other.text
            and self.items == other.items
        )

    def __hash__(self) -> int:
        return hash((self.kind, self.text))


# ---------------------------------------------------------------------------
# Flatten / Nicen / Indent helpers
# ---------------------------------------------------------------------------

def _flatten(item: EnvelopeFormatItem) -> list[EnvelopeFormatItem]:
    if item.kind == FormatItemKind.LIST:
        result: list[EnvelopeFormatItem] = []
        for child in item.items:
            result.extend(_flatten(child))
        return result
    return [item]


def _nicen(items: list[EnvelopeFormatItem]) -> list[EnvelopeFormatItem]:
    inp = list(items)
    result: list[EnvelopeFormatItem] = []
    while inp:
        current = inp.pop(0)
        if not inp:
            result.append(current)
            break
        if (
            current.kind == FormatItemKind.END
            and inp[0].kind == FormatItemKind.BEGIN
        ):
            result.append(
                EnvelopeFormatItem.end(f"{current.text} {inp[0].text}")
            )
            result.append(EnvelopeFormatItem.begin(""))
            inp.pop(0)
        else:
            result.append(current)
    return result


def _indent(level: int) -> str:
    return " " * (level * 4)


def _add_space_at_end_if_needed(s: str) -> str:
    if not s:
        return " "
    if s.endswith(" "):
        return s
    return s + " "


# ---------------------------------------------------------------------------
# Format flat / hierarchical
# ---------------------------------------------------------------------------

def _format_flat(item: EnvelopeFormatItem) -> str:
    line = ""
    for it in _flatten(item):
        if it.kind == FormatItemKind.BEGIN:
            if not line.endswith(" "):
                line += " "
            line += it.text
            line += " "
        elif it.kind == FormatItemKind.END:
            if not line.endswith(" "):
                line += " "
            line += it.text
            line += " "
        elif it.kind == FormatItemKind.ITEM:
            line += it.text
        elif it.kind == FormatItemKind.SEPARATOR:
            line = line.rstrip() + ", "
        elif it.kind == FormatItemKind.LIST:
            for child in it.items:
                line += _format_flat(child)
    return line


def _format_hierarchical(item: EnvelopeFormatItem) -> str:
    lines: list[str] = []
    level = 0
    current_line = ""

    for it in _nicen(_flatten(item)):
        if it.kind == FormatItemKind.BEGIN:
            if it.text:
                if not current_line:
                    c = it.text
                else:
                    c = _add_space_at_end_if_needed(current_line) + it.text
                lines.append(_indent(level) + c + "\n")
            level += 1
            current_line = ""
        elif it.kind == FormatItemKind.END:
            if current_line:
                lines.append(_indent(level) + current_line + "\n")
                current_line = ""
            level -= 1
            lines.append(_indent(level) + it.text + "\n")
        elif it.kind == FormatItemKind.ITEM:
            current_line += it.text
        elif it.kind == FormatItemKind.SEPARATOR:
            if current_line:
                lines.append(_indent(level) + current_line + "\n")
                current_line = ""
        elif it.kind == FormatItemKind.LIST:
            lines.append("<list>")

    if current_line:
        lines.append(current_line)

    return "".join(lines)


# ---------------------------------------------------------------------------
# Per-type format_item builders
# ---------------------------------------------------------------------------

def _cbor_format_item(
    cbor: CBOR,
    flat: bool,
    context: FormatContextOpt,
) -> EnvelopeFormatItem:
    """Format a CBOR value for envelope notation."""
    from ._cbor import envelope_tags, from_untagged_cbor

    # Check if this is a tagged envelope (tag 200)
    if cbor.case == CBORCase.TAGGED:
        tag, inner = cbor.value
        tags = envelope_tags()
        if tags and tag.value == tags[0].value:
            try:
                env = from_untagged_cbor(inner)
                return _envelope_format_item(env, flat, context)
            except Exception:
                return EnvelopeFormatItem.item("<error>")

    try:
        text = cbor_envelope_summary(cbor, 2**63, context)
        return EnvelopeFormatItem.item(text)
    except Exception:
        return EnvelopeFormatItem.item("<error>")


def _known_value_format_item(
    kv: object,  # KnownValue
    flat: bool,
    context: FormatContextOpt,
) -> EnvelopeFormatItem:
    """Format a KnownValue for envelope notation."""
    from known_values import KnownValue as KV

    value: KV = kv  # type: ignore[assignment]

    if context.is_none:
        name = value.name
    elif context.is_global:
        def _lookup(ctx: object) -> str:
            from ._format_context import FormatContext
            fc: FormatContext = ctx  # type: ignore[assignment]
            assigned = fc.known_values.assigned_name_for(value)
            return assigned if assigned is not None else value.name
        name = with_format_context(_lookup)
    else:
        assigned = context.context.known_values.assigned_name_for(value)
        name = assigned if assigned is not None else value.name

    return EnvelopeFormatItem.item(flanked_by(name, "'", "'"))


def _assertion_format_item(
    assertion: Assertion,
    flat: bool,
    context: FormatContextOpt,
) -> EnvelopeFormatItem:
    """Format an Assertion for envelope notation."""
    return EnvelopeFormatItem.list_([
        _envelope_format_item(assertion.predicate, flat, context),
        EnvelopeFormatItem.item(": "),
        _envelope_format_item(assertion.object, flat, context),
    ])


def _envelope_format_item(
    envelope: Envelope,
    flat: bool,
    context: FormatContextOpt,
) -> EnvelopeFormatItem:
    """Build the format item tree for an envelope."""
    ct = envelope.case_type
    case = envelope.case

    if ct == CaseType.LEAF:
        return _cbor_format_item(case.cbor, flat, context)

    if ct == CaseType.WRAPPED:
        return EnvelopeFormatItem.list_([
            EnvelopeFormatItem.begin("{"),
            _envelope_format_item(case.envelope, flat, context),
            EnvelopeFormatItem.end("}"),
        ])

    if ct == CaseType.ASSERTION:
        return _assertion_format_item(case.assertion, flat, context)

    if ct == CaseType.KNOWN_VALUE:
        return _known_value_format_item(case.value, flat, context)

    if ct == CaseType.ENCRYPTED:
        return EnvelopeFormatItem.item("ENCRYPTED")

    if ct == CaseType.COMPRESSED:
        return EnvelopeFormatItem.item("COMPRESSED")

    if ct == CaseType.ELIDED:
        return EnvelopeFormatItem.item("ELIDED")

    if ct == CaseType.NODE:
        return _format_node(case.subject, case.assertions, flat, context)

    return EnvelopeFormatItem.item("<error>")  # pragma: no cover


def _format_node(
    subject: Envelope,
    assertions: list[Envelope],
    flat: bool,
    context: FormatContextOpt,
) -> EnvelopeFormatItem:
    """Build format items for a Node envelope."""
    subject_item = _envelope_format_item(subject, flat, context)

    elided_count = 0
    encrypted_count = 0
    compressed_count = 0
    type_assertion_items: list[list[EnvelopeFormatItem]] = []
    assertion_items: list[list[EnvelopeFormatItem]] = []

    for assertion_env in assertions:
        act = assertion_env.case_type
        if act == CaseType.ELIDED:
            elided_count += 1
        elif act == CaseType.ENCRYPTED:
            encrypted_count += 1
        elif act == CaseType.COMPRESSED:
            compressed_count += 1
        else:
            item = [_envelope_format_item(assertion_env, flat, context)]
            # Check if this is a type assertion (isA predicate)
            is_type_assertion = False
            pred_env = assertion_env.as_predicate()
            if pred_env is not None:
                kv = pred_env.subject().as_known_value()
                if kv is not None and kv == IS_A:
                    is_type_assertion = True

            if is_type_assertion:
                type_assertion_items.append(item)
            else:
                assertion_items.append(item)

    # Sort by format item comparison key
    def _group_sort_key(
        group: list[EnvelopeFormatItem],
    ) -> tuple:
        return tuple(i._sort_key() for i in group)

    type_assertion_items.sort(key=_group_sort_key)
    assertion_items.sort(key=_group_sort_key)

    # Prepend type assertions
    all_assertion_items = type_assertion_items + assertion_items

    # Add compressed/elided/encrypted markers (order matches Rust)
    if compressed_count > 1:
        all_assertion_items.append(
            [EnvelopeFormatItem.item(f"COMPRESSED ({compressed_count})")]
        )
    elif compressed_count > 0:
        all_assertion_items.append([EnvelopeFormatItem.item("COMPRESSED")])

    if elided_count > 1:
        all_assertion_items.append(
            [EnvelopeFormatItem.item(f"ELIDED ({elided_count})")]
        )
    elif elided_count > 0:
        all_assertion_items.append([EnvelopeFormatItem.item("ELIDED")])

    if encrypted_count > 1:
        all_assertion_items.append(
            [EnvelopeFormatItem.item(f"ENCRYPTED ({encrypted_count})")]
        )
    elif encrypted_count > 0:
        all_assertion_items.append([EnvelopeFormatItem.item("ENCRYPTED")])

    # Intersperse separators between assertion groups
    joined: list[EnvelopeFormatItem] = []
    for i, group in enumerate(all_assertion_items):
        if i > 0:
            joined.append(EnvelopeFormatItem.separator())
        joined.extend(group)

    # Build the result items
    items: list[EnvelopeFormatItem] = []
    needs_braces = subject.is_subject_assertion()

    if needs_braces:
        items.append(EnvelopeFormatItem.begin("{"))
    items.append(subject_item)
    if needs_braces:
        items.append(EnvelopeFormatItem.end("}"))
    items.append(EnvelopeFormatItem.begin("["))
    items.extend(joined)
    items.append(EnvelopeFormatItem.end("]"))

    return EnvelopeFormatItem.list_(items)


# ---------------------------------------------------------------------------
# Public formatting entry points
# ---------------------------------------------------------------------------

def format_envelope(
    envelope: Envelope,
    *,
    flat: bool = False,
    context: FormatContextOpt | None = None,
) -> str:
    """Return the envelope notation string for an envelope.

    Parameters
    ----------
    envelope:
        The envelope to format.
    flat:
        If ``True``, format on a single line.
    context:
        Format context to use.  Defaults to global.
    """
    if context is None:
        context = FormatContextOpt.global_()
    item = _envelope_format_item(envelope, flat, context)
    if flat:
        return _format_flat(item).strip()
    return _format_hierarchical(item).strip()
